src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -6425,5 +7198,251 @@ public static class Extensions
                 Diagnostic(ErrorCode.ERR_UseDefViolation, "c").WithArguments("c").WithLocation(9, 33)
                 );
         }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaValidExtensionInClosestNamespaceInvalidInFurtherNamespace()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+using N1;
+using N3;
+
+namespace N1
+{
+    using N2;
+    public class C
+    {
+        public static async Task Main()
+        {
+            await foreach (var i in new C())
+            {
+                Console.Write(i);
+            }
+        }
+        public sealed class Enumerator
+        {
+            public int Current { get; private set; }
+            public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+        }
+    }
+}
+
+namespace N2
+{
+    public static class Extensions
+    {
+        public static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+    }
+}
+
+namespace N3
+{
+    public static class Extensions
+    {
+        public static int GetAsyncEnumerator(this C self) => throw null;
+    }
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (5,1): hidden CS8019: Unnecessary using directive.
+                // using N3;
+                Diagnostic(ErrorCode.HDN_UnusedUsingDirective, "using N3;").WithLocation(5, 1));
+            CompileAndVerify(comp, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaInvalidExtensionInClosestNamespaceValidInFurtherNamespace()
```

Comment:
The way these two tests are structured is somewhat confusing because the distance is based on the `using`. I would rather have it be structured such that `N1 { N2 { N3 { C } } }`, where the valid and invalid extensions are in N1 and N2.

Reply:
I'll add tests for that too.